junit-vintage-engine/src/main/java/org/junit/vintage/engine/VintageTestEngine.java

```diff
@@ -92,32 +95,46 @@ public void execute(ExecutionRequest request) {
 
 	private void executeAllChildren(VintageEngineDescriptor engineDescriptor,
 			EngineExecutionListener engineExecutionListener, ExecutionRequest request) {
-		boolean parallelExecutionEnabled = getParallelExecutionEnabled(request);
+		initializeParallelExecutionParameters(request);
 
-		if (parallelExecutionEnabled) {
-			if (executeInParallel(engineDescriptor, engineExecutionListener, request)) {
-				Thread.currentThread().interrupt();
-			}
+		boolean parallelExecutionEnabled = getParallelExecutionEnabled(request);
+		if (!parallelExecutionEnabled) {
+			executeSequentially(engineDescriptor, engineExecutionListener);
+			return;
 		}
-		else {
+
+		if (!classes && !methods) {
+			logger.warn(() -> "Parallel execution is enabled but no scope is defined. "
+					+ "Falling back to sequential execution.");
 			executeSequentially(engineDescriptor, engineExecutionListener);
+			return;
+		}
+
+		if (executeInParallel(engineDescriptor, engineExecutionListener, request)) {
+			Thread.currentThread().interrupt();
 		}
 	}
 
 	private boolean executeInParallel(VintageEngineDescriptor engineDescriptor,
 			EngineExecutionListener engineExecutionListener, ExecutionRequest request) {
-		ExecutorService executorService = Executors.newFixedThreadPool(getThreadPoolSize(request));
+		ExecutorService executorService = Executors.newWorkStealingPool(getThreadPoolSize(request));
 		RunnerExecutor runnerExecutor = new RunnerExecutor(engineExecutionListener);
 
-		List<CompletableFuture<Void>> futures = new ArrayList<>();
-		for (Iterator<TestDescriptor> iterator = engineDescriptor.getModifiableChildren().iterator(); iterator.hasNext();) {
-			TestDescriptor descriptor = iterator.next();
-			CompletableFuture<Void> future = CompletableFuture.runAsync(() -> {
-				runnerExecutor.execute((RunnerTestDescriptor) descriptor);
-			}, executorService);
+		List<RunnerTestDescriptor> runnerTestDescriptors = collectRunnerTestDescriptors(engineDescriptor,
+			executorService);
 
+		if (!classes) {
+			for (RunnerTestDescriptor runnerTestDescriptor : runnerTestDescriptors) {
+				runnerExecutor.execute(runnerTestDescriptor);
+			}
```

Comment:
Extract a method for readability?

```suggestion
			executeClassesSequentially(runnerTestDescriptors, runnerExecutor);
```